This is a patch to create the customized version of SWIG used for pywin32.
It was created by running the following command on MINGW64 (in git bash):
`diff --unified --recursive --ignore-space-change SWIG1.1p5-orig/ SWIG1.1p5/ > .../pywin32/SWIG/pywin32_swig.patch`

The patch is against SWIG-1.1p5 which can be found at:

  https://sourceforge.net/projects/swig/files/swig/1.1p5/

or

  https://web.archive.org/web/20110527063221/http://starship.python.net/crew/mhammond/downloads/swig1.1p5.tar.gz

To build the version of SWIG:

* Unpack the swig tarball
* Apply this patch (`git apply --ignore-space-change .../pywin32/SWIG/pywin32_swig.patch` from the root)
* Open a Developper Command Prompt/PowerShell in the "Win" directory
* Execute `nmake -f makefile.vc`
* Copy the newly created swig.exe from the swig tree to the pywin32 tree.


Huge thanks to Amaury Forgeot d'Arc and Roger Upole for rescuing this!

diff --unified --recursive --ignore-space-change SWIG1.1p5-orig/Modules/makefile.msc SWIG1.1p5/Modules/makefile.msc
--- SWIG1.1p5-orig/Modules/makefile.msc	1997-08-26 22:21:50.000000000 -0400
+++ SWIG1.1p5/Modules/makefile.msc	2025-12-15 10:46:16.208067800 -0500
@@ -52,7 +52,7 @@
 TARGET  = ..\swig.exe
 #CFLAGS  = -g -O -DSWIG_LIB='"$(SWIG_LIB)"' -DSWIG_CC='"$(CC)"'  -DSTDC_HEADERS=1 -DHAVE_ALLOCA_H=1 -DHAVE_ALLOCA=1 -DSWIG_LANG=PYTHON -DSWIG_DOC=ASCII -DHAVE_LIBDL=1 
 CFLAGS   = -Zi -nologo -DSWIG_LIB="\"$(SWIG_LIB)\"" -DSWIG_CC="\"$(CC)\""  -DMSDOS -DSTDC_HEADERS=1 -DHAVE_LIBDL=1 $(SWIG_OPTS)
-INCLUDE = -I../Include -I../SWIG -I$(STD_INC)
+INCFLAGS = -I../Include -I../SWIG
 LIBNAME = ..\libswig.lib
 LIBS    = $(LIBNAME)
 
@@ -60,12 +60,12 @@
 # Rules for creation of a .obj file from .cxx
 .SUFFIXES: .cxx
 .cxx.obj: 
-	$(CC) $(INCLUDE) $(CFLAGS) -c -o $*.obj $<
+	$(CC) $(INCFLAGS) $(CFLAGS) -c -o $*.obj $<
 
 all:    $(TARGET) 
 
 $(TARGET): $(WRAPOBJS) $(WRAPHEADERS) $(LIBNAME)
-	$(CC)  -o $(TARGET) $(INCLUDE) $(WRAPOBJS) $(LIBS)
+	$(CC)  -o $(TARGET) $(INCFLAGS) $(WRAPOBJS) $(LIBS)
 
 swigmain.obj: swigmain.cxx 
 tcl.obj: tcl.cxx 
diff --unified --recursive --ignore-space-change SWIG1.1p5-orig/Modules/python.cxx SWIG1.1p5/Modules/python.cxx
--- SWIG1.1p5-orig/Modules/python.cxx	1998-01-03 01:17:40.000000000 -0500
+++ SWIG1.1p5/Modules/python.cxx	2025-12-11 00:03:32.740350600 -0500
@@ -61,6 +61,8 @@
 
 static   int          doc_index = 0;
 static   DocString   *doc_strings = 0;
+static   int          PythonCom = 0; // Generate PyWin32 header files
+static   char        *PythonCom_Parent = "IUnknown";
 
 static char *usage = "\
 Python Options (available with -python)\n\
@@ -113,6 +115,14 @@
           } else if (strcmp(argv[i],"-docstring") == 0) {
 	    docstring = 1;
 	    mark_arg(i);
+          } else if (strcmp(argv[i],"-pythoncom") == 0) {
+	    PythonCom = 1;
+	    mark_arg(i);
+          } else if (strcmp(argv[i],"-com_interface_parent") == 0) {
+	    mark_arg(i);
+	    PythonCom_Parent = copy_string(argv[i+1]);
+	    mark_arg(i+1);
+            i++;
 	  } else if (strcmp(argv[i],"-help") == 0) {
 	    fputs(usage,stderr);
 	  }
@@ -258,6 +268,47 @@
   }
   fprintf(f_wrappers,"\t { NULL, NULL }\n");
   fprintf(f_wrappers,"};\n");
+  if (PythonCom) {
+    fprintf(f_wrappers,"PyComTypeObject Py%s::type(\"Py%s\",", module, module);
+    fprintf(f_wrappers,"\t&Py%s::type,\n", PythonCom_Parent);
+    fprintf(f_wrappers,"\tsizeof(Py%s),\n", module);
+    fprintf(f_wrappers,"\t%sMethods,\n", module);
+    fprintf(f_wrappers,"\tGET_PYCOM_CTOR(Py%s));", module);
+  }
+}
+
+// ---------------------------------------------------------------------
+// PYTHON::print_com_header()
+//
+// Prints out the COM header file
+// ---------------------------------------------------------------------
+
+void PYTHON::print_com_header() {
+  char fn_header[256];
+  sprintf(fn_header, "Py%s.h", module);
+  FILE *f_header = fopen(fn_header, "w");
+
+  if (!f_header) {
+    fprintf(stderr, "Unable to open %s\n", fn_header);
+    SWIG_exit(1);
+  }
+
+  fprintf(f_header, "class Py%s : public Py%s\n", module, PythonCom_Parent);
+  fprintf(f_header, "{\npublic:\nMAKE_PYCOM_CTOR(Py%s);\n", module);
+  fprintf(f_header, "static PyComTypeObject type;\nstatic %s *GetI(PyObject *self);\n", module);
+
+  Method *n = head;
+
+  n = head;
+  while (n) {
+    fprintf(f_header, "\tstatic PyObject *%s(PyObject *self, PyObject *args);\n", n->name);
+    n = n->next;
+  }
+
+  fprintf(f_header, "protected:\n\tPy%s(IUnknown *);\n\t~Py%s();\n", module, module);
+  fprintf(f_header, "};\n\n");
+
+  fclose(f_header);
 }
 
 // ---------------------------------------------------------------------
@@ -376,6 +427,8 @@
 void PYTHON::initialize_cmodule(void)
 {
   int i;
+  if (PythonCom)
+    return;
   fprintf(f_header,"#define SWIG_init    init%s\n\n", module);
   fprintf(f_header,"#define SWIG_name    \"%s\"\n", module);	
 
@@ -389,7 +442,7 @@
   fprintf(f_init,"extern \"C\" \n");
   fprintf(f_init,"#endif\n");
 
-  fprintf(f_init,"SWIGEXPORT(void,init%s)() {\n",module);
+  fprintf(f_init,"SWIGEXPORT(PyObject*, PyInit_%s)(void) {\n",module);
   fprintf(f_init,"\t PyObject *m, *d;\n");
 
   if (InitNames) {
@@ -400,8 +453,21 @@
     }
   }
   fprintf(f_init,"\t SWIG_globals = SWIG_newvarlink();\n");
-  fprintf(f_init,"\t m = Py_InitModule(\"%s\", %sMethods);\n", module, module);
-  fprintf(f_init,"\t d = PyModule_GetDict(m);\n");
+  fprintf(f_init,
+    "  static PyModuleDef %s_def = {\n"
+    "    PyModuleDef_HEAD_INIT,\n"
+    "      \"%s\",\n"
+    "      \"\",\n"
+    "      -1,\n"
+    "      %sMethods,\n"
+    "      };\n"
+    "  m = PyModule_Create(&%s_def);\n"
+    "  if (!m)\n"
+    "    return NULL;\n"
+    "  d = PyModule_GetDict(m);\n"
+    "  if (!d)\n"
+    "    return NULL;\n",
+    module, module, module, module);
 }
 
 
@@ -417,6 +483,9 @@
 
   print_methods();
   close_cmodule();
+  if (PythonCom) {
+    print_com_header();
+  }
   if ((doc_entry) && (module)){
     String temp;
     temp << "Python Module : ";
@@ -469,7 +538,11 @@
 // --------------------------------------------------------------------
 void PYTHON::close_cmodule(void)
 {
-  emit_ptr_equivalence(f_init);
+  if (PythonCom)
+    return;
+  // pywin32 doesn't need the pointer type-equivalency mappings.
+  // emit_ptr_equivalence(f_init);
+  fprintf(f_init,"\treturn m;\n");
   fprintf(f_init,"}\n");
 }
 
@@ -598,8 +671,11 @@
 // ----------------------------------------------------------------------
 void PYTHON::emit_function_header(WrapperFunction &emit_to, char *wname)
 {
-  emit_to.def << "static PyObject *" << wname
+  if (!PythonCom)
+    emit_to.def << "static ";
+  emit_to.def << "PyObject *" << wname
 	      << "(PyObject *self, PyObject *args) {";
+  if (!PythonCom)
   emit_to.code << tab4 << "self = self;\n";
 }
 
@@ -612,8 +688,13 @@
 //
 // Returns the name of the variable to use as the self pointer
 // ----------------------------------------------------------------------
-char *PYTHON::convert_self(WrapperFunction &)
+char *PYTHON::convert_self(WrapperFunction &f)
 {
+  if (PythonCom) {
+    f.code << "\t" << module << " *_swig_self;\n";
+    f.code << "\t" << "if ((_swig_self=GetI(self))==NULL) return NULL;\n";
+    return "_swig_self->";
+  }
   // Default behaviour is no translation
   return "";
 }
@@ -625,7 +706,14 @@
 // ----------------------------------------------------------------------
 char *PYTHON::make_funcname_wrapper(char *fnName)
 {
+  if (PythonCom) {
+    String wrapper;
+    wrapper << "Py" << module << "::" << fnName;
+    return wrapper;
+  }
+  else {
   return name_wrapper(fnName,"");
+  }
 }
 
 // ----------------------------------------------------------------------
@@ -1629,7 +1717,13 @@
 // -----------------------------------------------------------------------
 
 void PYTHON::add_native(char *name, char *funcname) {
+  if (PythonCom) {
+    String method;
+    method << "Py" << module << "::" << funcname;
+    add_method(name, method);
+  } else {
   add_method(name, funcname);
+  }
   if (shadow) {
     func << name << " = " << module << "." << name << "\n\n";
   }
diff --unified --recursive --ignore-space-change SWIG1.1p5-orig/Modules/python.h SWIG1.1p5/Modules/python.h
--- SWIG1.1p5-orig/Modules/python.h	1997-07-25 01:18:50.000000000 -0400
+++ SWIG1.1p5/Modules/python.h	2025-12-10 23:58:15.505195300 -0500
@@ -140,6 +140,7 @@
   // Add for Python-COM support
   virtual void initialize_cmodule();
   virtual void close_cmodule();
+  virtual void print_com_header();
   virtual void emit_function_header(WrapperFunction &emit_to, char *wname);
   virtual char *convert_self(WrapperFunction &f);
   virtual char *make_funcname_wrapper(char *fnName);
diff --unified --recursive --ignore-space-change SWIG1.1p5-orig/SWIG/makefile.msc SWIG1.1p5/SWIG/makefile.msc
--- SWIG1.1p5-orig/SWIG/makefile.msc	1997-06-23 18:15:32.000000000 -0400
+++ SWIG1.1p5/SWIG/makefile.msc	2025-12-15 10:46:30.109624300 -0500
@@ -42,7 +42,7 @@
 
 LIBHEADERS = internal.h ../Include/swig.h latex.h ascii.h html.h nodoc.h
 LIBNAME      = ..\libswig.lib
-INCLUDE  = -I../Include -I$(STD_INC)
+INCFLAGS  = -I../Include
 CFLAGS   = -Zi -nologo -DSWIG_LIB="\"$(SWIG_LIB)\"" -DSWIG_CC="\"$(CC)\""  -DMSDOS -DSTDC_HEADERS=1 -DHAVE_LIBDL=1 $(SWIG_OPTS)
 LD_FLAGS    =  -VERBOSE
 
@@ -51,7 +51,7 @@
 # Rules for creation of a .obj file from .cxx
 .SUFFIXES: .cxx
 .cxx.obj:
-	$(CC) $(INCLUDE) $(CFLAGS) -c -o $*.obj $<
+	$(CC) $(INCFLAGS) $(CFLAGS) -c -o $*.obj $<
 
 all:    $(LIBNAME) 
 
@@ -60,7 +60,7 @@
 	@$(LD) $(LD_FLAGS) -out:$(LIBNAME) $(LIBOBJS) parser.obj 
 
 parser.obj: parser.cxx $(LIBHEADERS)
-	$(CC) $(INCLUDE) $(CFLAGS) parser.cxx -c -o parser.obj
+	$(CC) $(INCFLAGS) $(CFLAGS) parser.cxx -c -o parser.obj
 
 parser.cxx: $(PARSER)
 	@echo "Must rebuild the parser with yacc"
